File: web/src/api/level.ts
import { request } from './client';

/**
 * Уровень сербского и оценка сложности текста.
 *
 * Уровень живёт на аккаунте, а не в разделе: спросили один раз — знают везде.
 * До этого о нём знал только Вукоток и хранил его при устройстве, поэтому один
 * человек отвечал на один и тот же вопрос в каждом браузере заново.
 */

export const LEVELS = ['A1', 'A2', 'B1', 'B2', 'C1'] as const;
export type Level = (typeof LEVELS)[number];

/** Как называется ступень для человека — «B1» само по себе ничего не говорит. */
export const LEVEL_NAMES: Record<Level, string> = {
  A1: 'Первые слова',
  A2: 'Простые фразы',
  B1: 'Читаю с переводчиком',
  B2: 'Читаю почти свободно',
  C1: 'Свободно',
};

export interface LevelQuestion {
  id: string;
  level: Level;
  prompt: string;
  hint: string;
  options: string[];
}

export interface LevelTestResult {
  level: Level;
  correct: number;
  total: number;
  byLevel: Record<string, number>;
}

export function getLevelTest() {
  return request<{ questions: LevelQuestion[] }>('/v1/profile/level/test', {
    anonymous: true,
  }).then((response) => response.questions);
}

/**
 * Ответы проверяет сервер.
 *
 * Верных вариантов у клиента нет и быть не должно: лежи они в исходниках
 * страницы, тест перестал бы что-либо измерять.
 */
export function gradeLevelTest(answers: Record<string, number>, save: boolean) {
  return request<LevelTestResult>('/v1/profile/level/test', {
    method: 'POST',
    body: { answers, save },
  });
}

export interface TextLevel {
  /** Пусто, если слов не хватило для суждения. */
  level: string;
  words: number;
  coverage: number;
  hardWords: string[];
  source: string;
}

/**
 * Оценивает, на какой уровень рассчитан текст.
 *
 * Отправляется выборка, а не книга целиком: оценщик всё равно смотрит не больше
 * 1200 слов, и гнать роман через сеть ради этого незачем.
 */
export function estimateTextLevel(paragraphs: string[], signal?: AbortSignal) {
  return request<TextLevel>('/v1/analyze/text-level', {
    method: 'POST',
    body: { paragraphs: sampleParagraphs(paragraphs) },
    anonymous: true,
    signal,
  });
}

/** Сколько абзацев отправлять: с запасом на короткие. */
const SAMPLE_PARAGRAPHS = 120;

/**
 * Берёт абзацы равномерно по всей книге.
 *
 * Именно по всей: начало книги врёт чаще всего — у переводных изданий там
 * выходные данные, у учебников предисловие на другом языке. Шаг постоянный, а
 * первый и последний абзацы берутся всегда.
 */
export function sampleParagraphs(paragraphs: string[]): string[] {
  const meaningful = paragraphs.filter((item) => item.trim().length > 0);
  if (meaningful.length <= SAMPLE_PARAGRAPHS) return meaningful;
  const step = meaningful.length / SAMPLE_PARAGRAPHS;
  const out: string[] = [];
  for (let i = 0; i < SAMPLE_PARAGRAPHS; i += 1) {
    const paragraph = meaningful[Math.floor(i * step)];
    if (paragraph !== undefined) out.push(paragraph);
  }
  return out;
}

/**
 * Разрыв, при котором стоит предупредить о книге.
 *
 * Две ступени, а не одна: читать на ступень выше своего уровня как раз и
 * полезно, и отговаривать от этого значит мешать единственному способу вырасти.
 * Правило то же, что на сервере (level.TooHardFor) — здесь оно нужно, чтобы не
 * ходить на сервер второй раз за уже известным ответом.
 */
export function tooHardFor(textLevel: string, readerLevel: string): boolean {
  const text = LEVELS.indexOf(textLevel as Level);
  const reader = LEVELS.indexOf(readerLevel as Level);
  return text >= 0 && reader >= 0 && text - reader >= 2;
}

File: frontend/lib/models/level.dart
/// Шкала CEFR в порядке возрастания.
const serbianLevels = <String>['A1', 'A2', 'B1', 'B2', 'C1'];

/// Как называется ступень для человека: «B1» само по себе ничего не говорит.
const serbianLevelNames = <String, String>{
  'A1': 'Первые слова',
  'A2': 'Простые фразы',
  'B1': 'Читаю с переводчиком',
  'B2': 'Читаю почти свободно',
  'C1': 'Свободно',
};

/// Один вопрос теста на уровень. Верного ответа здесь нет: проверяет сервер.
class LevelQuestion {
  final String id;
  final String level;

  /// Предложение с пропуском. Пропуск обозначен «___».
  final String prompt;

  /// Перевод на русский. Без него вопрос проверял бы знание слов, а не
  /// грамматики: не поняв фразы, наугад отвечают все одинаково.
  final String hint;
  final List<String> options;

  const LevelQuestion({
    required this.id,
    required this.level,
    required this.prompt,
    required this.hint,
    required this.options,
  });

  factory LevelQuestion.fromJson(Map<String, dynamic> j) => LevelQuestion(
        id: (j['id'] ?? '').toString(),
        level: (j['level'] ?? '').toString(),
        prompt: (j['prompt'] ?? '').toString(),
        hint: (j['hint'] ?? '').toString(),
        options: [
          for (final o in (j['options'] as List?) ?? const []) o.toString(),
        ],
      );
}

class LevelTestResult {
  final String level;
  final int correct;
  final int total;

  const LevelTestResult({
    required this.level,
    required this.correct,
    required this.total,
  });

  factory LevelTestResult.fromJson(Map<String, dynamic> j) => LevelTestResult(
        level: (j['level'] ?? 'A1').toString(),
        correct: (j['correct'] as num?)?.toInt() ?? 0,
        total: (j['total'] as num?)?.toInt() ?? 0,
      );
}

/// Насколько труден текст.
class TextLevel {
  /// Пусто, если слов не хватило для суждения.
  final String level;
  final int words;

  /// Самые редкие слова текста. Показываются человеку: «книга трудная» без
  /// примеров звучит как приговор без объяснения.
  final List<String> hardWords;

  const TextLevel({
    required this.level,
    required this.words,
    required this.hardWords,
  });

  factory TextLevel.fromJson(Map<String, dynamic> j) => TextLevel(
        level: (j['level'] ?? '').toString(),
        words: (j['words'] as num?)?.toInt() ?? 0,
        hardWords: [
          for (final w in (j['hardWords'] as List?) ?? const []) w.toString(),
        ],
      );
}

/// Стоит ли предупредить читателя о книге.
///
/// Разрыв в две ступени, а не в одну: читать на ступень выше своего уровня как
/// раз и полезно, и отговаривать от этого значит мешать единственному способу
/// вырасти. Правило то же, что на сервере (level.TooHardFor).
bool tooHardFor(String textLevel, String readerLevel) {
  final text = serbianLevels.indexOf(textLevel);
  final reader = serbianLevels.indexOf(readerLevel);
  return text >= 0 && reader >= 0 && text - reader >= 2;
}
